### ISSUE TITLE
Unexpected interactions between unions and type parameters

### ISSUE THREAD

**Issue by brentyi** [microsoft/pyright#9047]
Hi!

Not sure if this is a bug, but wanted to report some behavior I found surprising.

Here's a snippet that uses a type parameter. As expected, Pyright raises an error on the function call:

```python
def func[T](x: tuple[T, list[T]]) -> None: ...

list_of_int: list[int] = [3]

x = func(
    (None, list_of_int)  # Error
)
```
([pyright playground](https://pyright-play.net/?code=CYUwZgBGCuB2DGBtAKgXQBQA8BcEAu0ADgDYgoA0ExAlgM54qqoCUEAtAHwQByA9rCFwA6EQChRNegH1eYKdVh5ckhgryoIAXgiIAzKnGYtUOPHSiIliOj4DKKmXLWsIAYggBRAE5feX0cxAA))

However, if I modify the argument annotation with any kind of union, the error disappears:

```python
def func[T](x: tuple[T, list[T]] | str) -> None: ...

list_of_int: list[int] = [3]

x = func(
    (None, list_of_int)  # No error?
)
```

```python
def func[T](x: tuple[T, list[T]] | list[T]) -> None: ...

list_of_int: list[int] = [3]

x = func(
    (None, list_of_int)  # No error?
)
```

I verified this in the Pyright playground, and locally on v1.1.381. Is this expected?

Thanks!

**Comment by erictraut** [comment]
I agree this is a bug. Thanks for reporting it. I'll investigate further.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.382.
